### DISCUSSION TITLE
variable objective in different state

### DISCUSSION THREAD

**Discussion by ice-bird** [do-mpc/do-mpc#195] [Q&A]
I am trying to get the vehicle to avoid obstacles。there are two method，but method2 throws an error：
### method1
```python
d_obt=1000000
for vehicle in resource[:,t,:]:
    if (vehicle==[0.,0.]).any():
        d0=1000000
    else:
        d0 = sqrt((v_x-vehicle[0])**2+(v_y-vehicle[1])**2)
    d_obt = d_obt if d_obt<d0 else d0
    
if(d_obt<=dmin):
      mterm=1000000
  else:
      mterm=d_obt

mpc.set_objective(mterm=mterm,lterm=lterm)
```
in this method, vehicle is a point ,if distance between two vehicles small than  required min distance, which means collision, so i set mterm a large value.

### method2
```python
  v1 = Polygon(retPoint).convex_hull

  me=1
  # 从时空资源矩阵中寻找存在的车辆
  for vehicle in resource[:,t,:]:
      if (vehicle==[0.,0.,0.]).any():
          #no vehicle ,so it's safe
          me=1
      else:
          # retPoint2 stores 4 Vertex coordinates, and build a rectangle by shapely package
          retPoint2=np.zeros(shape=(4,2))
          v2 = Polygon(retPoint2).convex_hull

          if not v1.intersects(v2):
              # whether overlap？
              me=1
          else:
               # Overlap means collision
              me=10000000
              break

  if (me>1):
      mterm=1
  else:
      mterm=10000000

mpc.set_objective(mterm=mterm,lterm=lterm)

```
in this method, vehicle is a rectangle ,if two vehicle rectangle overlao, which means collision, so i set mterm a large value.

ERROR :

> in set_objective assert mterm.shape == (1,1), 'mterm must have shape=(1,1). You have {}'.format(mterm.shape)
> AttributeError: 'float' object has no attribute 'shape'

How to achieve variable objective？ thanks

**Comment by ice-bird** [comment]
in method2 I modified the code，but program throws the same error
```python
    if (me==1):
        mterm=1
    else:
        mterm=10000000
```

**Comment by 4flixt** [comment]
Hi @ice-bird,

So I am not entirely sure from your code snippet what is supposed to happen. But it seems like you want the cost function ``mterm`` to change conditionally. In do-mpc this term must be a “static” expression that only depends on variables that were previously introduced in the the ``model``. Often, you can achieve complex behavior (time-changing cost functions) with our [time-varying parameter](https://www.do-mpc.com/en/latest/FAQ.html#time-varying-parameters) implementation.

Best
4flixt
